### DISCUSSION TITLE
Project config parameters description

### DISCUSSION THREAD

**Discussion by j3k0** [airsdk/apm#59] [Ideas]
Packages parameters are defined with an ID and a default value. A nice addition would be some text description.

Currently:

```sh
$ apm project config
apsEnvironment=development
facebookAutoInitEnabled=true
```

My idealized output:

```sh
$ apm project config

# Apple push notifications environment (development | production)
apsEnvironment=development

# Tell the Facebook SDK not to initialize automatically
# See https://docs.airnativeextensions.com/docs/facebookapi/core/initialise-the-extension
facebookAutoInitEnabled=true
```

--

In `package.json` this could be done like so:

```json
{
  "parameters": [
    {
      "name": "facebookAutoInitEnabled",
      "defaultValue": "true",
      "description": "Tell the Facebook SDK not to initialize automatically\nSee https://docs.airnativeextensions.com/docs/facebookapi/core/initialise-the-extension"
    }
  ]
}

**Comment by j3k0** [comment]
Two additional ideas:

- <s>Mark some parameters as `"required": true`, so it breaks the build when they are not set. Useful when there's no sensible default, like the `facebookAppId` for instance.</s> (already exists apparently)
- List the extensions using a given parameter, so it makes it easier to go look for the related documentation.

**Reply by marchbold** [reply]
The required flag already exists and should work as you described. But I do plan to do some more work around it. I was thinking required values could be presented when you install a package as a question, (which can be skipped). Just to make a user aware that the package had configuration options required. Currently they can slip past a little too easily I feel? 

We like the description addition, feel free to log that as enhancement issue request.  We may need some logic around it's handling, eg if there are multiple packages using the same value, we probably don't want to build up an essay of duplicate descriptions so might have to just choose the latest?  Or maybe it could be more like:

```
$ apm project config describe <param>

<param>=XXX

Description:

[com.package.a] : Package A description

[com.package.b] : Package B description 
```

**Reply by j3k0** [reply]
I created an issue here, however it wouldn't let me change the `label` to set it as an enhancement.

**Reply by j3k0** [reply]
> I was thinking required values could be presented when you install a package as a question, (which can be skipped). Just to make a user aware that the package had configuration options required. Currently they can slip past a little too easily I feel?

...or just displaying the list of new parameters (for those not already set in the project config), with those required shown in bold font or something.

But ideally, I think `apm generate app-descriptor` should just fail with an error if a required parameter is not set. Required parameters should probably have empty default value.

That's the expected workflow:

1. install a plugin
2. configure parameters
3. generate app-descriptor

So any errors should probably be caught in step 3.
